### PR TITLE
Feature exception handling

### PR DESCRIPTION
**Pull request by FFroehlich**

### CODE REVIEW

**Comment by FFroehlich** [comment]
not all return codes are removed yet, and apparently some of them were not doing what they were supposed to (steadyStateProblem and Event code). But let's generate a coverage report first to see where we are still missing tests.

File: src/newton_solver.cpp [hunk 1]
```diff
@@ -298,70 +252,61 @@ NewtonSolverSparse::NewtonSolverSparse(Model *model, ReturnData *rdata,
...
+        numeric = klu_factor((tdata->J)->indexptrs, (tdata->J)->indexvals,
+                             (tdata->J)->data, symbolic, &common);
+        if (numeric) {
+            return;
         } else {
-            status = AMICI_ERROR_NEWTON_LINSOLVER;
+            throw NewtonFailure("KLU factorization failed!");
         }
+    } else {
+        throw NewtonFailure("KLU symbolic analisis failed!");
```
Comment (dweindl) [inline src/newton_solver.cpp]: "analysis"

File: src/newton_solver.cpp [hunk 2]
```diff
@@ -410,38 +368,150 @@ int NewtonSolverIterative::prepareLinearSystem(int ntry, int nnewt) {
...
+    N_VDestroy_Serial(ns_h);
+    N_VDestroy_Serial(ns_t);
+    N_VDestroy_Serial(ns_s);
+    N_VDestroy_Serial(ns_r);
+    N_VDestroy_Serial(ns_rt);
+    N_VDestroy_Serial(ns_v);
+    N_VDestroy_Serial(ns_Jv);
+    N_VDestroy_Serial(ns_tmp);
+    N_VDestroy_Serial(ns_Jdiag);
+    N_VScale(-1.0, tdata->xdot, tdata->xdot);
```
Comment (dweindl) [inline src/newton_solver.cpp]: This is asking for trouble... Depending on when the object is destroyed, the scaling might happen at the wrong time.

File: src/rdata.cpp [hunk 3]
```diff
@@ -81,49 +109,56 @@ void ReturnData::setLikelihoodSensitivitySecondOrderNaN() {
...
-        return AMICI_SUCCESS;
+        return;
 
     // chain-rule factor: multiplier for am_p
     realtype coefficient;
-    realtype *pcoefficient, *augcoefficient;
 
-    pcoefficient = new realtype[nplist]();
-    augcoefficient = new realtype[np]();
+    std::vector<realtype> pcoefficient;
```
Comment (dweindl) [inline src/rdata.cpp]: Set size at construction (also below)

File: src/steadystateproblem.cpp [hunk 4]
```diff
@@ -32,66 +31,56 @@ int SteadystateProblem::workSteadyStateProblem(const UserData *udata,
...
-                status =
-                    solver->AMISensReInit(udata->ism, tdata->sx, tdata->sdx);
-        if (status != AMICI_SUCCESS)
-            status = AMICI_ERROR_PREEQUILIBRATION;
+                solver->AMISensReInit(udata->ism, tdata->sx, tdata->sdx);
     }
-
-    delete newtonSolver;
-    return status;
+    return;
```
Comment (dweindl) [inline src/steadystateproblem.cpp]: Superfluous return

File: src/udata.cpp [hunk 5]
```diff
@@ -76,7 +76,7 @@ UserData::UserData(const UserData &other) : UserData(other.np, other.nk, other.n
     }
 
     if(other.sx0data) {
-        sx0data = new double[other.nx];
+        sx0data = new double[other.nx * other.nplist];
```
Comment (dweindl) [inline src/udata.cpp]: :+1:

File: tests/cpputest/unittests/tests1.cpp [hunk 6]
```diff
@@ -27,15 +27,25 @@ TEST_GROUP(amici)
 TEST(amici, testRunAmiciSimulationStateMismatch) {
     UserData udata(1, 2, 3);
     Model model(4, 0, 3, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, AMICI_O2MODE_NONE);
-
-    CHECK_EQUAL(AMICI_ERROR_UDATA, runAmiciSimulation(&udata, nullptr, nullptr, &model));
+    bool exceptionThrown = false;
```
Comment (dweindl) [inline tests/cpputest/unittests/tests1.cpp]: `CHECK_THROWS(amici::AmiException,  runAmiciSimulation(&udata, nullptr, nullptr, &model));`

File: tests/cpputest/unittests/tests1.cpp [hunk 7]
```diff
@@ -27,15 +27,25 @@ TEST_GROUP(amici)
...
+    }
+    CHECK_TRUE(exceptionThrown)
 }
 
 TEST(amici, testRunAmiciSimulationRdataMissing) {
     UserData udata(1, 2, 3);
     Model model(1, 3, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, AMICI_O2MODE_NONE);
-
-    CHECK_EQUAL(AMICI_ERROR_RDATA, runAmiciSimulation(&udata, nullptr, nullptr, &model));
+    bool exceptionThrown = false;
```
Comment (dweindl) [inline tests/cpputest/unittests/tests1.cpp]: see above

File: tests/cpputest/unittests/tests1.cpp [hunk 8]
```diff
@@ -27,15 +27,25 @@ TEST_GROUP(amici)
...
+    try {
```
Comment (dweindl) [inline tests/cpputest/unittests/tests1.cpp]: see above

File: src/newton_solver.cpp [hunk 9]
```diff
@@ -410,38 +368,150 @@ int NewtonSolverIterative::prepareLinearSystem(int ntry, int nnewt) {
...
-     * @return stats integer flag indicating success of the method
      */
 
     newton_try = ntry;
     i_newton = nnewt;
     if (nnewt == -1) {
-        return AMICI_ERROR_NEWTON_LINSOLVER;
-    } else {
-        return AMICI_SUCCESS;
+        throw NewtonFailure("No setup necessary");
```
Comment (paulstapor) [inline src/newton_solver.cpp]: Here, the error message should be
throw NewtonFailure("Linear solver SPBCG does not support sensitivity computation for steady state problems");

File: src/newton_solver.cpp [hunk 10]
```diff
@@ -216,45 +177,38 @@ NewtonSolverDense::NewtonSolverDense(Model *model, ReturnData *rdata,
...
-
-    if (status == AMICI_SUCCESS) {
-        /* Compute factorization and pivoting */
-        status = DenseGETRF(tdata->Jtmp, pivots);
-    }
-
-    return status;
+    int status = DenseGETRF(tdata->Jtmp, pivots);
+    if(status != 0)
+        throw NewtonFailure("Dense factorization failed!");
```
Comment (paulstapor) [inline src/newton_solver.cpp]: I like this way to throw errors: Here, you can give much more detailed error messages.

**Pull request merged**